Getting Started with the Buf CLI (#355)
bufbuild/docs.buf.build: docs/tutorials/getting-started-with-buf-cli.md

---
id: getting-started-with-buf-cli
title: Getting Started with the Buf CLI
description: "Hello Buf CLI! Goodbye manual code generation tasks. let's get started and empower your development
workflow!"
---

The Buf CLI is the ultimate tool for modern, fast, and efficient Protobuf API management. With features such as
formatting, linting, breaking change detection and code generation, Buf offers a comprehensive solution for Protobuf
development and maintenance. Buf is designed to integrate seamlessly with your existing workflow, so you can focus on
what matters most: writing great APIs. Whether you are working with a small, focused project or a massive, complex
system, Buf is the perfect choice. In the next 10 minutes, you will learn how to use the Buf CLI to easily build, lint,
format and generate code for your project.

:::info

We will assume you have already installed [`buf`](/installation)

> and [`git`](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git) in your `$PATH`. If you haven't, head on over

Reviewer: Section 5 requires Go as well.

you would when specifying a [linting] configuration. Your `buf.yaml` file currently has the `FILE` option configured:

```yaml title="buf.yaml" {2-4}
version: v1
breaking:
  use:
    - FILE
lint:
  use:
    - DEFAULT
  ignore:
    - google/type/datetime.proto
```

### 4.1 Break Your API {#break-your-api}

Next, you'll need to introduce a breaking change. First, make a change that's breaking at the [`WIRE`][wire] level. This
is the most fundamental type of breaking change as it changes how the Protobuf messages are encoded in transit ("on the
wire"). This type of breaking change affects _all_ users in _all_ languages.

For example, change the type of the `Pet.pet_type` field from `PetType` to `string`:

```protobuf title=pet/v1/pet.proto {2-3}
 message Pet {
    -  PetType pet_type = 1;
    +  string pet_type = 1;
  string pet_id = 2;
  string name = 3;
}
```

### 4.2 Run `buf breaking` {#run-buf-breaking}

Now, verify that this is a breaking change against the local `main` branch. You'll also notice errors related to
the changes you made in the [previous step](#lint-your-api):

```terminal

> buf breaking proto --against "../../.git#branch=main,subdir=start/getting-started-with-buf-cli/proto"

Reviewer: With this still in dev, the `branch=main` doesn't work, but removing the branch altogether works (and should work even if the user changes branches, but not if they commit):
```suggestion
buf breaking proto --against "../../.git#subdir=start/getting-started-with-buf-cli/proto"
```

---
proto/pet/v1/pet.proto:1:1:Previously present service "PetStore" was deleted from file.
proto/pet/v1/pet.proto:35:3:Field "1" on message "PutPetRequest" changed type from "enum" to "string".
proto/pet/v1/pet.proto:44:3:Field "1" with name "pet_id" on message "DeletePetRequest" changed option "json_name" from "petID" to "petId".
proto/pet/v1/pet.proto:44:10:Field "1" on message "DeletePetRequest" changed name from "petID" to "pet_id".
```

> For remote locations that require authentication, see
> [HTTPSAuthentication](../reference/inputs.md#https) and
> [SSH Authentication](../reference/inputs.md#ssh) for more details.

### 4.3 Revert changes {#revert-changes}

Once you've determined that your change is breaking, revert it:

```protobuf title=pet/v1/pet.proto {2-3}
 message Pet {
    -  string pet_type = 1;
    +  PetType pet_type = 1;
  string pet_id = 2;
  string name = 3;
}
```

## 5 Implement an API

In this section, you'll implement a `PetStoreService` client and server, both of which you can run on the command line.

```terminal

> cd ..

Reviewer: what directory should I be in? Before this command, I'm in `buf-tour/start/getting-started-with-buf-cli`. Running the next few steps from `buf-tour/start` breaks on resolving deps.

Reviewer: Get the following when I make the go module in `getting-started-with-buf-cli`. (Similar but different errors if it lives under start)

```
github.com/bufbuild/buf-tour/petstore/gen/pet/v1 imports
	github.com/bufbuild/buf-tour/start/getting-started-with-buf-cli/gen/google/type: module github.com/bufbuild/buf-tour@latest found (v0.0.0-20230118150351-e3df1294ecfe), but does not contain package github.com/bufbuild/buf-tour/start/getting-started-with-buf-cli/gen/google/type
github.com/bufbuild/buf-tour/petstore/gen/pet/v1/petv1connect imports
	github.com/bufbuild/buf-tour/start/getting-started-with-buf-cli/gen/pet/v1: module github.com/bufbuild/buf-tour@latest found (v0.0.0-20230118150351-e3df1294ecfe), but does not contain package github.com/bufbuild/buf-tour/start/getting-started-with-buf-cli/gen/pet/v1
```

Reviewer: Just for simplicity, I'd say let's not get fancy with rewriting the module's import path and just have it match the repo as-is (likewise for the gen config).

	"github.com/bufbuild/buf-tour/petstore/gen/pet/v1/petv1connect"
	"github.com/bufbuild/connect-go"
	"golang.org/x/net/http2"
	"golang.org/x/net/http2/h2c"
)
const address = "localhost:8080"
func main() {
	mux := http.NewServeMux()
	path, handler := petv1connect.NewPetStoreServiceHandler(&petStoreServiceServer{})
	mux.Handle(path, handler)
	fmt.Println("... Listening on", address)
	http.ListenAndServe(
		address,
		// Use h2c so we can serve HTTP/2 without TLS.
		h2c.NewHandler(mux, &http2.Server{}),
	)
}
// petStoreServiceServer implements the PetStoreService API.
type petStoreServiceServer struct {
	petv1connect.UnimplementedPetStoreServiceHandler
}
// PutPet adds the pet associated with the given request into the PetStore.
func (s *petStoreServiceServer) PutPet(
	ctx context.Context,
	req *connect.Request[petv1.PutPetRequest],
) (*connect.Response[petv1.PutPetResponse], error) {
	name := req.Msg.GetName()
	petType := req.Msg.GetPetType()
	log.Println("Got a request to create a", petType, "named", name)
	return connect.NewResponse(&petv1.PutPetResponse{}), nil
}
```

### 5.3 Resolve Go dependencies {#resolve-go-dependencies}

> Now that you have code for both a client and a server, run this command to resolve some of the dependencies you need for
> the generated code:

Reviewer: ```suggestion
Now that you have code for a server, run this command to resolve the dependencies you need to build the code:
```

```terminal
go mod tidy
```

### 5.4 Call `PutPet` {#call-putpet}

> With the `server/main.go` and `client/main.go` implementations shown above, run

Reviewer: ```suggestion
With the `server/main.go` implementation shown above, run
```

> the server and call the `PutPet` endpoint from the client.

Reviewer: ```suggestion
the server and call the `PutPet` endpoint from the buf CLI.
```

First, run the server:

```terminal
go run server/main.go
---
... Listening on 127.0.0.1:8080
```

> In a separate terminal, hit the API with a curl command:

Reviewer: ```suggestion
In a separate terminal, hit the API with a `buf curl` command:
```
